assembly/src/parsers/io_ops.rs

```diff
@@ -256,69 +276,145 @@ pub fn parse_loc_storew(op: &Token) -> Result<Node, ParsingError> {
+        ));
+    }
+
+    if param_str.len() > HEX_CHUNK_SIZE {
+        return Err(ParsingError::invalid_param_with_reason(
+            op,
+            param_idx,
+            &format!("hex string '{param_str}' contains too many characters"),
+        ));
     }
 
-    Ok(constants)
+    let value = u64::from_str_radix(param_str, 16)
+        .map_err(|_| ParsingError::invalid_param(op, param_idx))?;
+
+    if value >= Felt::MODULUS {
+        Err(ParsingError::invalid_param_with_reason(
+            op,
+            param_idx,
+            &format!("hex string '{param_str}' contains value greater than field modulus"),
+        ))
+    } else {
+        Ok(value)
+    }
 }
 
-fn parse_hex_params(
-    op: &Token,
-    param_idx: usize,
-    param_str: &str,
-) -> Result<Vec<Felt>, ParsingError> {
-    // handle error cases where the hex string is poorly formed
-    let is_single_element = if param_str.len() <= HEX_CHUNK_SIZE {
-        if param_str.len() % 2 != 0 {
-            // parameter string is not a valid hex representation
-            return Err(ParsingError::invalid_param(op, param_idx));
-        }
-        true
+/// Determines the minimal type appropriate for provided value and returns appropriate instruction
+/// for this value
+fn build_push_one_instruction(value: u64) -> Result<Node, ParsingError> {
+    if value <= u8::MAX as u64 {
+        Ok(Instruction(PushU8(value as u8)))
+    } else if value <= u16::MAX as u64 {
+        Ok(Instruction(PushU16(value as u16)))
+    } else if value <= u32::MAX as u64 {
+        Ok(Instruction(PushU32(value as u32)))
+    } else if value < Felt::MODULUS {
+        Ok(Instruction(PushFelt(Felt::new(value))))
     } else {
-        if param_str.len() % HEX_CHUNK_SIZE != 0 {
-            // hex string doesn't contain a valid number of bytes
-            return Err(ParsingError::invalid_param(op, param_idx));
-        }
-        false
-    };
+        unreachable!()
+    }
+}
 
-    // parse the hex string into one or more valid field elements
-    if is_single_element {
-        // parse a single element in hex representation
-        let parsed_param = parse_hex_param(op, param_idx, param_str)?;
-        Ok(vec![parsed_param])
+/// Determines the minimal type appropriate for provided values iterator and returns appropriate
+/// instruction for this values
+fn build_push_many_instruction<I>(values_iter: I) -> Result<Node, ParsingError>
+where
+    I: Iterator<Item = Result<u64, ParsingError>> + Clone + ExactSizeIterator,
+{
+    assert!(values_iter.len() != 0);
+    let max_value = values_iter
+        .clone()
+        .try_fold(0, |max, value| value.map(|value| value.max(max)))?;
+    // let values_max = values_iter.clone().map(Result::unwrap).max();
+    if max_value <= u8::MAX as u64 {
+        let values_u8 = values_iter
+            .map(|v| v.map(|v| v as u8))
+            .collect::<Result<Vec<u8>, _>>()?;
+        Ok(Instruction(PushU8List(values_u8)))
```

Comment:
For clarity, it might be better to create a separate function. Something like:
```Rust
fn cast_u64<T>(value: Result<u64, ParsingError>) -> Result<T, ParsingError>
where
    T: TryFrom<u64> + Debug,
    <T as TryFrom<u64>>::Error: Debug,
{
    value.map(|value| value.try_into().expect("out of bounds"))
}
```
And then this line (and other lines) becomes something like this:
```Rust
let values_u8 = values.map(cast_u64).collect::<Result<Vec<u8>, _>>()?;
Ok(Instruction(PushU8List(values_u8)))
```
@vlopes11 - what do you think?

Reply:
To clarify: would you go with `cast_u64()` function or would you keep the `.map(|v| v.map(|v| v as u8))` inline?

Reply:
`cast_u64` won't bring a clear gain because we can't coerce in generics anyway. A map of a map on the other hand is too complicated and usually can be improved.

I would offer a third alternative that is to simplify the expression into something like `values_iter.clone().try_fold(0, |max, value| Ok(value?.max(max)))`

Reply:
> `cast_u64` won't bring a clear gain because we can't coerce in generics anyway.

The way I have it above it actually works generically. Or do you mean something else?

Reply:
@bobbinth we end up adding more instructions if we specify how we want to panic (with the `try_into().expect("message")`).

The native coerce is the cheapest option so its better to always stick to it, whenever viable.

I wrote a small example here. We can see that `foo` ends up with more instructions than `bar`

https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:22,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'use+std::convert::%7BTryInto,+TryFrom%7D%3B%0Ause+std::fmt::Debug%3B%0A%0Apub+type+ParsingError+%3D+%26!'static+str%3B%0A%0Apub+fn+foo(value:+u64)+-%3E+Result%3Cu32,+ParsingError%3E+%7B%0A++++cast_u64(Ok(value))%0A%7D%0A%0Apub+fn+bar(value:+u64)+-%3E+Result%3Cu32,+ParsingError%3E+%7B%0A++++Ok(value+as+u32)%0A%7D%0A%0Apub+fn+cast_u64%3CT%3E(value:+Result%3Cu64,+ParsingError%3E)+-%3E+Result%3CT,+ParsingError%3E%0Awhere%0A++++T:+TryFrom%3Cu64%3E+%2B+Debug,%0A++++%3CT+as+TryFrom%3Cu64%3E%3E::Error:+Debug,%0A%7B%0A++++value.map(%7Cvalue%7C+value.try_into().expect(%22out+of+bounds%22))%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.333333333333336,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1570,deviceViewOpen:'1',filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:22,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.57.0+(Editor+%231)',t:'0')),header:(),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'rustc+1.57.0',editorid:1,fontScale:22,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+1.57.0+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

Reply:
yep - makes sense!

Reply:
And I guess for casting we can do it like this:
```Rust
let values_u8 = values.map(|v| Ok(v? as u8)).collect::<Result<Vec<u8>, _>>()?;
```

Reply:
That looks good! I hope someday in the future the language designers will improve the experience with iterators of results